File: ce326/hw2/YUVImage.java

package ce326.hw2;

import java.io.BufferedReader;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.io.PrintWriter;
import static java.lang.Character.isWhitespace;
import java.util.Scanner;
import ce326.hw2.YUVPixel;

public class YUVImage {
    YUVPixel[][] array;
    int width;
    int height;
    
    
    public YUVImage(int width, int height){
        this.width = width;
        this.height = height;
        
        array = new YUVPixel[height][width];
        
        for(int i= 0; i < height; i++){
            for(int j = 0; j < width; j++){
                array[i][j] = new YUVPixel((short)16, (short)128, (short)128);
            }
        }
    }
    
    public YUVImage(YUVImage copyImg){
        this.width = copyImg.width;
        this.height = copyImg.height;
        
        array = new YUVPixel[height][width];
        
        for(int i= 0; i < height; i++){
            for(int j = 0; j < width; j++){
                array[i][j] = new YUVPixel(copyImg.array[i][j].getY(), copyImg.array[i][j].getU(), copyImg.array[i][j].getV());
            }
        }
    }
    
    public YUVImage(RGBImage RGBImg){
        this.width = RGBImg.width;
        this.height = RGBImg.height;
        
        array = new YUVPixel[height][width];
        
        for(int i= 0; i < height; i++){
            for(int j = 0; j < width; j++){
                array[i][j] = new YUVPixel(RGBImg.array[i][j]);
            }
        }
    }
    
    public YUVImage(java.io.File file){
        int times = 0;//mas leei se poia metavliti anaferetai h kathe timi pou yparxei sto arxeio
        int first = 1;//distinguish height, width, colordepth from red, blue, green
        int heightHelp = 0;
        int widthHelp = 0;
        String strPpm;
        
        try{
            strPpm = file.getName();
            if(!((strPpm.charAt(strPpm.length() - 1) == 'v') & (strPpm.charAt(strPpm.length() - 2) == 'u') & (strPpm.charAt(strPpm.length() - 3) == 'y') & (strPpm.charAt(strPpm.length() - 4) == '.'))){
                throw new UnsupportedFileFormatException();
            }
            
            Scanner readerFile = new Scanner(file);
            readerFile.next();
            
            while(readerFile.hasNext()){
                String nextString = readerFile.next();
                times++;
                if(first == 1){
                    if(times == 1){
                        width = Integer.parseInt(nextString);
                    }
                    else if(times == 2){
                        height = Integer.parseInt(nextString);

                        array = new YUVPixel[height][width];
                        for(int k = 0; k < height; k++){//initialisation rgb array
                            for(int j = 0; j < width; j++){
                                array[k][j] = new YUVPixel();
                            }
                        }

                        first = 0;
                        times = 0;
                    }
                }
                else{
                    if(times == 1){
                        array[heightHelp][widthHelp].setY(Short.valueOf(nextString));
                    }
                    else if(times == 2){
                        array[heightHelp][widthHelp].setU(Short.valueOf(nextString));
                    }
                    else if(times == 3){
                        array[heightHelp][widthHelp].setV(Short.valueOf(nextString));
                        times = 0;
                        if((heightHelp == height-1) & (widthHelp == width-1)){
                            break;
                        }
                        else if(widthHelp == (width-1)){//reinitialisation
                            widthHelp = 0;
                            heightHelp++;
                        }
                        else{
                            widthHelp++;
                        }
                    } 

                }
                

            }
            readerFile.close();
            
        }
        catch(FileNotFoundException e){}
        catch(UnsupportedFileFormatException ex){} 
    }
    
    @Override
    public String toString(){
        String str = "";
        StringBuilder strbuilderHelp = new StringBuilder(str);
        strbuilderHelp.insert(strbuilderHelp.length(), "YUV3");
        strbuilderHelp.insert(strbuilderHelp.length(), String.format(" %d %d", width, height));
        
        for(int i = 0; i < height; i++){
            for(int j = 0; j < width; j++){
                strbuilderHelp.insert(strbuilderHelp.length(), String.format(" %d %d %d", array[i][j].getY(), array[i][j].getU(), array[i][j].getV()));
            }
        }  
       
        str = String.valueOf(strbuilderHelp);
        System.out.println(str);

        return(str);
    }
    
    
    public void toFile(java.io.File file) {
        PrintWriter pw = null;
        
        try{
            
            if(file.exists()){//if file already exists -> empty
                PrintWriter print_writer = new PrintWriter(file);
                print_writer.print("");
                print_writer.close();
            }
            
            FileWriter writer = new FileWriter(file, true);
            pw = new PrintWriter(file);
            pw.println("YUV3" + "\n" + width + " " + height);
            for(int i = 0; i < height; i++){
                for(int j = 0; j <width; j++){
                    pw.println(array[i][j].getY() + " " + array[i][j].getU() + " " + array[i][j].getV());
                }
        }   
        
        }catch (IOException e) {
            System.out.println("An error occurred during reading");
        }finally {
            if (pw != null) {
               pw.close();
            }
        }
        
    } 
    
    public void equalize(){
        Histogram histogram = new Histogram(this);
        histogram.equalize();
        short value;
        
        //copy in a yuv array the balanced one
        for(int i = 0; i < this.height; i++){
            for(int j = 0; j < this.width; j++){
                value = this.array[i][j].getY();
                this.array[i][j].setY((short) histogram.balancedHistogramArray[value]);
            }
        }
    }
    
    
    
}

File: ce326/hw2/python/YUVImage.py
from RGBPixel import RGBPixel
from RGBImage import RGBImage
from YUVPixel import YUVPixel
from PPMImage import PPMImage
from Histogram import Histogram
from UnsupportedFileFormatException import UnsupportedFileFormatException
import os
import numpy as np
import os.path
import io
import imghdr
from pathlib import Path


class YUVImage:
    
    def __init__(self, width, height, imgorfile):
        self.width = width;
        self.height = height;
        self.array = np.empty(shape = (self.height, self.width), dtype = YUVPixel)
        
        #if the imgorfile variable is not an image nor a file 
        #in this case we just initiate with (16, 16, 16, 128)
        if((not isinstance(imgorfile, io.IOBase)) & (not(isinstance(imgorfile, YUVImage))) & (not(isinstance(imgorfile, RGBImage)))):
            for i in range(self.height):
                for j in range(self.width):
                    self.array[i][j] = YUVPixel(16, 16, 16, 128);
        #case that we want to copy the image from another YUVImage
        elif(isinstance(imgorfile, YUVImage)):
            self.width = imgorfile.width;
            self.height = imgorfile.height;
            
            for i in range(height):
                for j in range(width):
                    self.array[i][j] = YUVPixel(imgorfile.array[i][j].getY(), imgorfile.array[i][j].getU(), imgorfile.array[i][j].getV(), 0);
        
        #case that we want to copy the image from another RGBImage
        elif(isinstance(imgorfile, RGBImage)):
            self.width = imgorfile.width;
            self.height = imgorfile.height;
            
            for i in range(height):
                for j in range(width):
                    self.array[i][j] = YUVPixel(5, 0, 0, imgorfile);
            
        #case that we want to create the image from a file
        elif(isinstance(imgorfile, io.IOBase)):
            times = 0; #mas leei se poia metavliti anaferetai h kathe timi pou yparxei sto arxeio
            first = 1; #distinguish height, width, colordepth from red, blue, green
            heightHelp = 0;
            widthHelp = 0;
            var = 0;
            
            try:
                strPpm = os.path.basename(imgorfile.name)
                if((strPpm[len(strPpm) - 4 : len(strPpm)]) != '.yuv'):
                    raise UnsupportedFileFormatException()
                
                string = imgorfile.read().rstrip('\n')
      
                splits = string.split(); #traverse first object(YUV)
                
                #read the information from the .yuv file and save them to array, width, height
                for split in splits:
                    var = var + 1
                    if(var == 1):
                        continue
                    
                    times = times + 1;
                    if(first == 1):
                        if(times == 1):
                            self.width = int(split);
                        elif(times == 2):
                            self.height = int(split);

                            self.array = np.empty(shape = (self.height, self.width), dtype = YUVPixel)
                            for k in range(self.height): #initialisation rgb array
                                for j in range(self.width):
                                    self.array[k][j] = YUVPixel(0, 0, 0, 0);
                                    
                            first = 0;
                            times = 0;
                            
                    else:
                        if(times == 1):
                            self.array[heightHelp][widthHelp].setY(int(split));
                        
                        elif(times == 2):
                            self.array[heightHelp][widthHelp].setU(int(split));
                            
                        elif(times == 3):
                            self.array[heightHelp][widthHelp].setV(int(split));
                            times = 0;
                            if((heightHelp == self.height - 1) & (widthHelp == self.width - 1)):
                                break;
                            elif(widthHelp == (self.width - 1)): #reinitialisation
                                widthHelp = 0;
                                heightHelp = heightHelp + 1;
                            else:
                                widthHelp = widthHelp + 1;
                            
            except FileNotFoundError:
                print("File not Found!")
            except UnsupportedFileFormatException:
                print("UnsupportedFileFormatException")
            
    
    def toString(self):
        mystr = "YUV3";
        mystr += '\n'
        mystr += str(self.width);
        mystr += " "
        mystr += str(self.height);
        mystr += "\n"
        
        
        for i in range(self.height):
            for j in range(self.width):
                mystr += str(self.array[i][j].getY())
                mystr += " "
                mystr += str(self.array[i][j].getU())
                mystr += " "
                mystr += str(self.array[i][j].getV());
                mystr += "\n"

        return(mystr);
    
    
    def toFile(self, file):
        
        try:
            file.truncate(0)
            
            mystr = self.toString();
            file.write(mystr)
        
        except IOError:
           print("An error occurred during reading");
    
    def equalize(self):
        histogram = Histogram(self);
        histogram.equalize();
        
        #copy in a yuv array the balanced one
        for i in range(self.height):
            for j in range(self.width):
                    value = self.array[i][j].getY();
                    self.array[i][j].setY(histogram.balancedHistogramArray[value]);

